synapse/storage/databases/main/state.py

```diff
@@ -430,6 +430,11 @@ def _update_state_for_partial_state_event_txn(
             updatevalues={"state_group": state_group},
         )
 
+        # the event may now be rejected where it was not before, or vice versa,
+        # in which case we need to update the rejected flags.
+        if bool(context.rejected) != (event.rejected_reason is not None):
+            self.mark_event_rejected_txn(txn, event.event_id, context.rejected)
```

Comment:
Struggling to fully appreciate the scenarios in which this happens, so I'm just going to ask you a question:

If an event becomes rejected (or becomes accepted), do we not need to worry about events which refer to it becoming accepted/rejected themselves?

For example, if I send a join with my invitation as an auth event.
If the invitation is rejected, I take it that my join will also be rejected.

If this invitation later becomes accepted, should that cascade down to my join?

If so, does your code take this into account (where?)?

(...This PR looks much too trivial for what it sounds like.)

Reply:
Yes, absolutely we do need to worry about that. 

But when we resync the state, we work through every event in the room (in the order we received them) re-evaluating the state at that event and whether we should have accepted it or not (we call [`update_state_for_partial_state_event`](https://github.com/matrix-org/synapse/blob/cf0d5d810c42218099b2496e09fccfbe5a398e38/synapse/handlers/federation_event.py#L521) for each one). So I think this should all come out in the wash.